---
layout: post
title: Helpful articles for quality control and pre-processing of RNA-seq data
date: '2023-06-05'
categories: [Bioinformatics]
tags: [FastQC, Sequencing, RNA-Seq]
---

1. Quality Control and Preprocessing of Sequencing Reads: https://bio-protocol.org/exchange/protocoldetail?id=4454&type=1

2. **FastQC**
- FastQC help analysis modules (how to interpret FastQC results): https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/

- two YouTube videos for walkthroughs of FastQC results: https://www.youtube.com/watch?v=lUk5Ju3vCDM and https://www.youtube.com/watch?v=bz93ReOv87Y

- Tutorial of how to intepret FastQC results: https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html

3. Why is 3'-Tag RNA-seq useful? https://dnatech.genomecenter.ucdavis.edu/faqs/when-do-you-recommend-3-tag-rna-seq/

4. What is QuantSeq 3' mRNA sequencing: https://www.nature.com/articles/nmeth.f.376

5. GC-content normalization for RNA-seq data: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-480

6. **CCS Student Mentors tutorials**
- https://github.com/ccsstudentmentors/tutorials/tree/master/RNA-Seq/Quantifying-RNA-Expression
<img width="1053" alt="Screen Shot 2023-06-05 at 9 25 20 AM" src="https://github.com/ademerlis/ademerlis.github.io/assets/56000927/9c0fc04e-047c-4c8d-bd56-140cfc01f7ae">
