---
layout: post
title: WGCNA for wound healing manuscript pt. 2
date: '2023-06-07'
categories: RNA, coding, woundhealingPdam
tags: [RNA, coding, woundhealingPdam]
---

I've been feeling like the heatmap I created doesn't quite make sense, because I feel like the only thing that should be on the x axis is condition x hour, not each individual like I have it currently:

<img width="1197" alt="Screen Shot 2023-06-06 at 10 59 58 AM" src="https://github.com/ademerlis/ademerlis.github.io/assets/56000927/a8407339-e316-4f06-98a8-52080e31c2d7">


One thing I found in this tutorial (https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html#42_Import_and_set_up_data) that could be important is that when they created the DESeqDataset for their example, they didn't specify any model because the purpose of the WGCNA is not to perform differential expression analysis. So maybe the fact that I specified a formula in the dds object (and more specifically, a complicated LRT one) might be affecting the comparisons.

Other links of interest:
1. https://wikis.utexas.edu/display/bioiteam/Clustering+using+WGCNA
2. https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-023-05233-z
3. https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04554-1
